packages/tpetra/core/src/Tpetra_Import_Util2.hpp

```diff
@@ -1237,28 +1250,263 @@ lowCommunicationMakeColMapAndReindexKokkos (const Teuchos::ArrayView<const size_
+      if(max < PIDList_view[i]) max = PIDList_view[i];
+    }, Kokkos::Max<int>(PID_max));
+  
+    using KeyViewTypePID = decltype(PIDList_view);
+    using BinSortOpPID = Kokkos::BinOp1D<KeyViewTypePID>;
+  
+    // Make a subview of ColIndices for remote GID sorting
+    auto ColIndices_subview = Kokkos::subview(ColIndices_view, Kokkos::make_pair(NumLocalColGIDs, ColIndices_view.size()));
+  
+    // Make binOp with bins = PID_max + 1, min = 0, max = PID_max
+    BinSortOpPID binOp2(PID_max+1, 0, PID_max);
+    
+    // Sort External column indices so that all columns coming from a
+    // given remote processor are contiguous.  This is a sort with one
+    // auxilary array: RemoteColIndices
+    Kokkos::BinSort<KeyViewTypePID, BinSortOpPID> bin_sort2(PIDList_view, 0, PIDList_view.size(), binOp2, false);
+    bin_sort2.create_permute_vector(exec);
+    bin_sort2.sort(exec, PIDList_view);
+    bin_sort2.sort(exec, ColIndices_subview);
+  
+    // Deep copy back from device to host
+    Kokkos::deep_copy(execution_space(), PIDList_host, PIDList_view);
+  
+    // Stash the RemotePIDs. Once remotePIDs is changed to become a Kokkos view, we can remove this and copy directly.
+    // Note: If Teuchos::Array had a shrink_to_fit like std::vector,
+    // we'd call it here.
+    
+    Teuchos::Array<int> PIDList(NumRemoteColGIDs);
+    for(LO i = 0; i < NumRemoteColGIDs; ++i) {
+      PIDList[i] = PIDList_host[i];
+    }
+  
+    remotePIDs = PIDList;
+  
+    // Sort external column indices so that columns from a given remote
+    // processor are not only contiguous but also in ascending
+    // order. NOTE: I don't know if the number of externals associated
+    // with a given remote processor is known at this point ... so I
+    // count them here.
+    LO StartCurrent = 0, StartNext = 1;
+    while(StartNext < NumRemoteColGIDs) {
+      if(PIDList_host[StartNext] == PIDList_host[StartNext-1]) {
+        StartNext++;
+      }
+      else {
+        Kokkos::sort(ColIndices_view, NumLocalColGIDs + StartCurrent, NumLocalColGIDs + StartNext);
+        StartCurrent = StartNext;
+        StartNext++;
+      }
+    }
+  
+    Kokkos::sort(ColIndices_view, NumLocalColGIDs + StartCurrent, NumLocalColGIDs + StartNext);
+  }
+
+
+  // Build permute array for *local* reindexing.
+
+  // Now fill front end. Two cases:
+  //
+  // (1) If the number of Local column GIDs is the same as the number
+  //     of Local domain GIDs, we can simply read the domain GIDs into
+  //     the front part of ColIndices, otherwise
+  //
+  // (2) We step through the GIDs of the domainMap, checking to see if
+  //     each domain GID is a column GID.  we want to do this to
+  //     maintain a consistent ordering of GIDs between the columns
+  //     and the domain.
+  Teuchos::ArrayView<const GO> domainGlobalElements = domainMap.getLocalElementList();
```

Comment:
You could get a device view of elements directly with ``domainMap.getMyGlobalIndices()``

Reply:
@brian-kelley Doesn't `Map::getMyGlobalIndices` return a host view?

https://github.com/trilinos/Trilinos/blob/22445c8febb4ba87bb6f2819e3b0ea8f4c79599b/packages/tpetra/core/src/Tpetra_Map_def.hpp#L1655

Reply:
Yes it does 🤦. ``Map`` and ``Details::LocalMap`` both carry around the device view we want (``lgMap_``), but it's private and there's no way to get it out. I guess it sort of makes sense since it's lazily populated by ``getMyGlobalIndices()``. I would suggest adding a version of ``Map::getMyGlobalIndices`` that does the same lazy stuff, but returns ``lgMap_`` instead of ``lgMapHost_``.

Reply:
BTW, I don't really like how Map's typedef for the host GID view is called ``global_indices_array_type``, without 'host' in the name. Since that typedef itself is private, we could probably change that easily.